nbqa/replace_source.py

```diff
@@ -87,24 +169,116 @@ def main(python_file: "Path", notebook: "Path", notebook_info: NotebookInfo) ->
     """
     notebook_json = json.loads(notebook.read_text())
 
-    pyfile = python_file.read_text()
+    pycells = _get_pycells(python_file)
 
-    pycells: Iterator[str] = iter(pyfile[len(CODE_SEPARATOR) :].split(CODE_SEPARATOR))
     code_cell_number = 0
     new_cells = []
     for cell in notebook_json["cells"]:
         if cell["cell_type"] == "code":
             code_cell_number += 1
-            if code_cell_number not in notebook_info.code_cells_to_ignore:
-                source = _restore_semicolon(
-                    next(pycells), code_cell_number, notebook_info.trailing_semicolons
-                )
-                cell["source"] = _reinstate_magics(
-                    source,
-                    notebook_info.temporary_lines.get(code_cell_number, []),
-                )
-
+            new_source = _get_new_source(code_cell_number, notebook_info, pycells)
```

Comment:
(redundant comment) This could also be refactored once the changes suggested for `diff` function gets implemented.

* generator to give only the code cells
*  `_get_new_source(code_cell_number, notebook_info, pycells)` becomes ` _get_new_source(code_cell_number, notebook_info, next(pycells))` when the for loop becomes
```Python
 for cell_no, cell in enumerate(_notebook_code_cells, start=1):
    if cell_no not in notebook_info.code_cells_to_ignore:
        #entire logic follows here
```